File: kreuzberg/packages/java/src/main/java/dev/kreuzberg/config/PaddleOcrConfig.java
package dev.kreuzberg.config;

import java.util.HashMap;
import java.util.Map;

/**
 * Configuration for PaddleOCR backend.
 *
 * <p>
 * Configures PaddleOCR-specific parameters including language, caching, text
 * detection, recognition, and preprocessing options.
 *
 * @since 4.4.0
 */
public final class PaddleOcrConfig {
	private final String language;
	private final String cacheDir;
	private final Boolean useAngleCls;
	private final Boolean enableTableDetection;
	private final Double detDbThresh;
	private final Double detDbBoxThresh;
	private final Double detDbUnclipRatio;
	private final Integer detLimitSideLen;
	private final Integer recBatchNum;

	private PaddleOcrConfig(Builder builder) {
		this.language = builder.language;
		this.cacheDir = builder.cacheDir;
		this.useAngleCls = builder.useAngleCls;
		this.enableTableDetection = builder.enableTableDetection;
		this.detDbThresh = builder.detDbThresh;
		this.detDbBoxThresh = builder.detDbBoxThresh;
		this.detDbUnclipRatio = builder.detDbUnclipRatio;
		this.detLimitSideLen = builder.detLimitSideLen;
		this.recBatchNum = builder.recBatchNum;
	}

	public static Builder builder() {
		return new Builder();
	}

	public String getLanguage() {
		return language;
	}

	public String getCacheDir() {
		return cacheDir;
	}

	public Boolean getUseAngleCls() {
		return useAngleCls;
	}

	public Boolean getEnableTableDetection() {
		return enableTableDetection;
	}

	public Double getDetDbThresh() {
		return detDbThresh;
	}

	public Double getDetDbBoxThresh() {
		return detDbBoxThresh;
	}

	public Double getDetDbUnclipRatio() {
		return detDbUnclipRatio;
	}

	public Integer getDetLimitSideLen() {
		return detLimitSideLen;
	}

	public Integer getRecBatchNum() {
		return recBatchNum;
	}

	public Map<String, Object> toMap() {
		Map<String, Object> map = new HashMap<>();
		if (language != null) {
			map.put("language", language);
		}
		if (cacheDir != null) {
			map.put("cache_dir", cacheDir);
		}
		if (useAngleCls != null) {
			map.put("use_angle_cls", useAngleCls);
		}
		if (enableTableDetection != null) {
			map.put("enable_table_detection", enableTableDetection);
		}
		if (detDbThresh != null) {
			map.put("det_db_thresh", detDbThresh);
		}
		if (detDbBoxThresh != null) {
			map.put("det_db_box_thresh", detDbBoxThresh);
		}
		if (detDbUnclipRatio != null) {
			map.put("det_db_unclip_ratio", detDbUnclipRatio);
		}
		if (detLimitSideLen != null) {
			map.put("det_limit_side_len", detLimitSideLen);
		}
		if (recBatchNum != null) {
			map.put("rec_batch_num", recBatchNum);
		}
		return map;
	}

	static PaddleOcrConfig fromMap(Map<String, Object> map) {
		if (map == null) {
			return null;
		}
		Builder builder = builder();
		if (map.get("language") instanceof String) {
			builder.language((String) map.get("language"));
		}
		if (map.get("cache_dir") instanceof String) {
			builder.cacheDir((String) map.get("cache_dir"));
		}
		if (map.get("use_angle_cls") instanceof Boolean) {
			builder.useAngleCls((Boolean) map.get("use_angle_cls"));
		}
		if (map.get("enable_table_detection") instanceof Boolean) {
			builder.enableTableDetection((Boolean) map.get("enable_table_detection"));
		}
		if (map.get("det_db_thresh") instanceof Number) {
			builder.detDbThresh(((Number) map.get("det_db_thresh")).doubleValue());
		}
		if (map.get("det_db_box_thresh") instanceof Number) {
			builder.detDbBoxThresh(((Number) map.get("det_db_box_thresh")).doubleValue());
		}
		if (map.get("det_db_unclip_ratio") instanceof Number) {
			builder.detDbUnclipRatio(((Number) map.get("det_db_unclip_ratio")).doubleValue());
		}
		if (map.get("det_limit_side_len") instanceof Number) {
			builder.detLimitSideLen(((Number) map.get("det_limit_side_len")).intValue());
		}
		if (map.get("rec_batch_num") instanceof Number) {
			builder.recBatchNum(((Number) map.get("rec_batch_num")).intValue());
		}
		return builder.build();
	}

	public static final class Builder {
		private String language;
		private String cacheDir;
		private Boolean useAngleCls;
		private Boolean enableTableDetection;
		private Double detDbThresh;
		private Double detDbBoxThresh;
		private Double detDbUnclipRatio;
		private Integer detLimitSideLen;
		private Integer recBatchNum;

		private Builder() {
		}

		public Builder language(String language) {
			this.language = language;
			return this;
		}

		public Builder cacheDir(String cacheDir) {
			this.cacheDir = cacheDir;
			return this;
		}

		public Builder useAngleCls(Boolean useAngleCls) {
			this.useAngleCls = useAngleCls;
			return this;
		}

		public Builder enableTableDetection(Boolean enableTableDetection) {
			this.enableTableDetection = enableTableDetection;
			return this;
		}

		public Builder detDbThresh(Double detDbThresh) {
			this.detDbThresh = detDbThresh;
			return this;
		}

		public Builder detDbBoxThresh(Double detDbBoxThresh) {
			this.detDbBoxThresh = detDbBoxThresh;
			return this;
		}

		public Builder detDbUnclipRatio(Double detDbUnclipRatio) {
			this.detDbUnclipRatio = detDbUnclipRatio;
			return this;
		}

		public Builder detLimitSideLen(Integer detLimitSideLen) {
			this.detLimitSideLen = detLimitSideLen;
			return this;
		}

		public Builder recBatchNum(Integer recBatchNum) {
			this.recBatchNum = recBatchNum;
			return this;
		}

		public PaddleOcrConfig build() {
			return new PaddleOcrConfig(this);
		}
	}
}

File: kreuzberg/packages/php/src/Config/PaddleOcrConfig.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Config;

/**
 * PaddleOCR-specific configuration.
 *
 * Advanced settings for the PaddleOCR engine including language selection,
 * model caching, angle detection, table recognition, and detection thresholds.
 */
readonly class PaddleOcrConfig
{
    public function __construct(
        /**
         * Language code for OCR recognition.
         *
         * Specifies which language model to use for text recognition.
         * Common codes: 'ch' (Chinese), 'en' (English), 'fr' (French), etc.
         *
         * @var string|null
         * @default null
         */
        public ?string $language = null,

        /**
         * Directory for caching PaddleOCR models.
         *
         * Models are downloaded and cached in this directory to avoid repeated
         * downloads. If null, uses PaddleOCR's default cache location.
         *
         * @var string|null
         * @default null
         */
        public ?string $cacheDir = null,

        /**
         * Enable angle classification for rotated text.
         *
         * When enabled, uses an additional angle classification model to detect
         * and correct rotated text before recognition. Improves accuracy on rotated
         * documents but requires additional model loading time.
         *
         * @var bool|null
         * @default null
         */
        public ?bool $useAngleCls = null,

        /**
         * Enable table structure detection.
         *
         * When enabled, uses table detection models to identify and parse table
         * structures within the document. Improves extraction from documents with
         * tabular data.
         *
         * @var bool|null
         * @default null
         */
        public ?bool $enableTableDetection = null,

        /**
         * Threshold for text region detection.
         *
         * Controls the sensitivity of text detection. Lower values detect more regions
         * (more false positives), higher values detect fewer regions (more false negatives).
         * Range: 0.0 to 1.0
         *
         * @var float|null
         * @default null
         */
        public ?float $detDbThresh = null,

        /**
         * Box filter threshold for text detection.
         *
         * Filters out detected text boxes with scores below this threshold.
         * Range: 0.0 to 1.0
         *
         * @var float|null
         * @default null
         */
        public ?float $detDbBoxThresh = null,

        /**
         * Unclip ratio for expanding detected text regions.
         *
         * Controls how much to expand detected text boxes. Higher values create
         * larger boxes. Typically between 1.0 and 2.5.
         *
         * @var float|null
         * @default null
         */
        public ?float $detDbUnclipRatio = null,

        /**
         * Limit side length for detection image resize.
         *
         * Maximum side length for resizing detection input. If an image dimension
         * exceeds this, the image is scaled down. Larger values improve accuracy
         * but require more memory and computation.
         *
         * @var int|null
         * @default null
         */
        public ?int $detLimitSideLen = null,

        /**
         * Batch size for recognition.
         *
         * Number of text regions to process simultaneously during recognition.
         * Higher values are faster but require more memory.
         *
         * @var int|null
         * @default null
         */
        public ?int $recBatchNum = null,
    ) {
    }

    /**
     * Create configuration from array data.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string|null $language */
        $language = $data['language'] ?? null;
        if ($language !== null && !is_string($language)) {
            /** @var string $language */
            $language = (string) $language;
        }

        /** @var string|null $cacheDir */
        $cacheDir = $data['cache_dir'] ?? null;
        if ($cacheDir !== null && !is_string($cacheDir)) {
            /** @var string $cacheDir */
            $cacheDir = (string) $cacheDir;
        }

        /** @var bool|null $useAngleCls */
        $useAngleCls = $data['use_angle_cls'] ?? null;
        if ($useAngleCls !== null && !is_bool($useAngleCls)) {
            /** @var bool $useAngleCls */
            $useAngleCls = (bool) $useAngleCls;
        }

        /** @var bool|null $enableTableDetection */
        $enableTableDetection = $data['enable_table_detection'] ?? null;
        if ($enableTableDetection !== null && !is_bool($enableTableDetection)) {
            /** @var bool $enableTableDetection */
            $enableTableDetection = (bool) $enableTableDetection;
        }

        /** @var float|null $detDbThresh */
        $detDbThresh = isset($data['det_db_thresh']) && is_numeric($data['det_db_thresh'])
            ? (float) $data['det_db_thresh']
            : null;

        /** @var float|null $detDbBoxThresh */
        $detDbBoxThresh = isset($data['det_db_box_thresh']) && is_numeric($data['det_db_box_thresh'])
            ? (float) $data['det_db_box_thresh']
            : null;

        /** @var float|null $detDbUnclipRatio */
        $detDbUnclipRatio = isset($data['det_db_unclip_ratio']) && is_numeric($data['det_db_unclip_ratio'])
            ? (float) $data['det_db_unclip_ratio']
            : null;

        /** @var int|null $detLimitSideLen */
        $detLimitSideLen = isset($data['det_limit_side_len']) && is_numeric($data['det_limit_side_len'])
            ? (int) $data['det_limit_side_len']
            : null;

        /** @var int|null $recBatchNum */
        $recBatchNum = isset($data['rec_batch_num']) && is_numeric($data['rec_batch_num'])
            ? (int) $data['rec_batch_num']
            : null;

        return new self(
            language: $language,
            cacheDir: $cacheDir,
            useAngleCls: $useAngleCls,
            enableTableDetection: $enableTableDetection,
            detDbThresh: $detDbThresh,
            detDbBoxThresh: $detDbBoxThresh,
            detDbUnclipRatio: $detDbUnclipRatio,
            detLimitSideLen: $detLimitSideLen,
            recBatchNum: $recBatchNum,
        );
    }

    /**
     * Create configuration from JSON string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode($json, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new \InvalidArgumentException('Invalid JSON: ' . json_last_error_msg());
        }
        if (!is_array($data)) {
            throw new \InvalidArgumentException('JSON must decode to an object/array');
        }
        /** @var array<string, mixed> $data */
        return self::fromArray($data);
    }

    /**
     * Create configuration from JSON file.
     */
    public static function fromFile(string $path): self
    {
        if (!file_exists($path)) {
            throw new \InvalidArgumentException("File not found: {$path}");
        }
        $contents = file_get_contents($path);
        if ($contents === false) {
            throw new \InvalidArgumentException("Unable to read file: {$path}");
        }
        return self::fromJson($contents);
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return array_filter([
            'language' => $this->language,
            'cache_dir' => $this->cacheDir,
            'use_angle_cls' => $this->useAngleCls,
            'enable_table_detection' => $this->enableTableDetection,
            'det_db_thresh' => $this->detDbThresh,
            'det_db_box_thresh' => $this->detDbBoxThresh,
            'det_db_unclip_ratio' => $this->detDbUnclipRatio,
            'det_limit_side_len' => $this->detLimitSideLen,
            'rec_batch_num' => $this->recBatchNum,
        ], static fn ($value): bool => $value !== null);
    }

    /**
     * Convert configuration to JSON string.
     */
    public function toJson(): string
    {
        $json = json_encode($this->toArray(), JSON_PRETTY_PRINT);
        if ($json === false) {
            throw new \RuntimeException('Failed to encode configuration to JSON');
        }
        return $json;
    }
}
